core-processor/src/main/java/io/micronaut/inject/beans/visitor/MapperVisitor.java

```diff
@@ -53,60 +53,91 @@ public void visitClass(ClassElement element, VisitorContext context) {
+            value.stringValue("to").ifPresent(to -> {
+                if (toDefs.contains(to)) {
+                    throw new ProcessingException(element, "Multiple @Mapping definitions map to the same property: " + to);
+                } else {
+                    toDefs.add(to);
+                    List<PropertyElement> beanProperties = toType.getBeanProperties(PropertyElementQuery.of(toType).includes(Set.of(to)));
+                    if (beanProperties.isEmpty()) {
+                        throw new ProcessingException(element, "@Mapping(to=\"" + to + "\") specifies a property that doesn't exist in type " + toType.getName());
+                    }
+                }
+            });
+            value.stringValue("from").ifPresent(from -> {
+                if (from.contains("#{")) {
+                    return;
+                }
+                if (from.contains(".")) {
+                    int index = from.indexOf(".");
+                    String argumentName = from.substring(0, index);
+                    String propertyName = from.substring(index + 1);
+
+                    boolean anyMatch = false;
+                    for (ParameterElement parameter: parameters) {
+                        if (parameter.getName().equals(argumentName)) {
+                            anyMatch = true;
+                            if (parameter.getType().getName().equals(Map.class.getName())) {
+                                break;
                             }
-                        }
-                        if (lastClassElement !=  null) {
-                            lastClassElement.annotate(Mapper.class);
+                            if (parameter.getGenericType().getBeanProperties().stream().noneMatch(p -> p.getName().equals(propertyName))) {
+                                throw new ProcessingException(element, "@Mapping(from=\"" + from + "\") specifies property " + propertyName + " that doesn't exist in type " + parameter.getGenericType().getName());
+                            }
+                            break;
                         }
                     }
+                    if (!anyMatch) {
+                        throw new ProcessingException(element, "@Mapping(from=\"" + from + "\") specifies argument " + argumentName + " that doesn't exist for method");
+                    }
                 } else {
-                    throw new ProcessingException(element, "Only a single parameter is permitted on a mapping method");
+                    String propertyName = from.substring(from.indexOf(".") + 1);
+                    for (ParameterElement parameter: parameters) {
+                        if (parameter.getType().getName().equals(Map.class.getName())) {
+                            continue;
+                        }
+                        if (parameter.getGenericType().getBeanProperties().stream().noneMatch(p -> p.getName().equals(propertyName))) {
```

Comment:
you can use `PropertyElementQuery` and use `includes` instead of querying all properties